moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java

```diff
@@ -285,14 +285,18 @@ public AdapterMethod(Type type, Set<? extends Annotation> annotations, Object ad
       this.nullable = nullable;
     }
 
-    public void bind(Moshi moshi) {
+    public void bind(Moshi moshi, JsonAdapter.Factory factory) {
       if (jsonAdapters.length > 0) {
         Type[] parameterTypes = method.getGenericParameterTypes();
         Annotation[][] parameterAnnotations = method.getParameterAnnotations();
         for (int i = adaptersOffset, size = parameterTypes.length; i < size; i++) {
-          jsonAdapters[i - adaptersOffset] = moshi.adapter(
-              ((ParameterizedType) parameterTypes[i]).getActualTypeArguments()[0],
-              jsonAnnotations(parameterAnnotations[i]));
+          Type type = ((ParameterizedType) parameterTypes[i]).getActualTypeArguments()[0];
+          Set<? extends Annotation> jsonAnnotations = jsonAnnotations(parameterAnnotations[i]);
+          if (this.type == type && annotations.equals(jsonAnnotations)) {
```

Comment:
this should probably be Types.equal(). In general neither `==` nor `.equals()` works on `java.lang.reflect.Type`. (for example, `String[].class` isn’t equal to the parameter of `List<String[]>.class`)